libStorage Token Support (JWT)

akutz opened the pull request:
This patch introduces support for JSON Web Tokens, used to restrict access to API resources to clients that provide valid tokens with sufficient claims. Please see the commit's doc section for additional information. The docs are also [built](http://libstorage-akutz.readthedocs.io/en/feature-token-support/user-guide/config/#authentication) on my fork at RTFD.

codenrhoden commented on .docs/user-guide/config.md:
````diff
@@ -451,6 +451,196 @@ that has signed the allowed client certificates. Or at the very least the
+```
+
+The above configuration snippet defines a new property,
+`libstorage.server.auth`. This property contains three child properties:
````
s/three/five/

codenrhoden commented on api/server/security/security_token.go:
```diff
@@ -0,0 +1,151 @@
+		allowed bool
+		denied  bool
+	)
+	for _, v := range config.Deny {
```
This code that loops through Denied and Allow subjects identical to that found in the service auth handler (when a previously validated token is already in the context). Could we move this bit of logic in a common function, such as `security.ValidateTokenSubject(config, tok.Subject)` and call that from both places?

codenrhoden commented on api/types/types_errors.go:
```diff
@@ -51,3 +53,34 @@ type ErrBatchProcess struct{ goof.Goof }
+
+// ErrMissingStorageService occurs when the storage service is expected in
+// the provided context but is not there.
+var ErrMissingStorageService = errors.New("missing storage service")
```
Why the use of `errors` instead of `goof` here? This one is defined differently than `ErroTimedOut` and `ErrNotImplemented`

akutz replied:
Hi @codenrhoden,

This has been handled.

codenrhoden reviewed (changes requested):
This is looking good. Apologies that two of my review comments went out as single comments rather than all at once in single review!

I know you noted this already, but just for posterity there is not any logic in place that would return an HTTP 401 error if you do an API call on multiple services w/o the JWT and one of the services requires it.

akutz commented:
Hi @codenrhoden,

FYI there *is* now logic to handle the multi-service auth. I created a new handler at [`handlers_auth_all_services.go`](https://github.com/akutz/libstorage/blob/feature/token-support/api/server/handlers/handlers_auth_all_services.go) that is applied to the following routes:

* `/services/`
* `/volumes/`
* `/snapshots/`

The latest update also handles the duplicate code we discussed. Thank you again for providing such a comprehensive review.

akutz committed:
libStorage Token Support (JWT)

This patch introduces support for JSON Web Tokens, used to restrict
access to API resources to clients that provide valid tokens with
sufficient claims.
